cloud-controller-manager/do/firewall_controller.go

```diff
@@ -0,0 +1,309 @@
+}
+
+// Run starts the firewall controller loop.
+func (fc *FirewallController) Run(stopCh <-chan struct{}) {
+	wait.Until(func() {
+		ctx, cancel := context.WithTimeout(context.Background(), serviceSyncPeriod)
+		defer cancel()
+
+		currentFirewall, err := fc.fwManager.Get(ctx)
+		if err != nil {
+			klog.Errorf("failed to get worker firewall: %s", err)
+		}
+
+		if currentFirewall != nil {
+			fc.fwCache.mu.RLock()
+			isEqual := cmp.Equal(currentFirewall, fc.fwCache.firewall)
+			fc.fwCache.mu.RUnlock()
+
+			if isEqual {
+				return
+			}
+		}
+		err = fc.ensureReconciledFirewall(ctx)
+		if err != nil {
+			klog.Errorf("failed to reconcile worker firewall: %s", err)
+		}
+
+	}, firewallReconcileRequency, stopCh)
+}
+
+// Get returns the current public access firewall representation.
+func (fm *firewallManagerOp) Get(ctx context.Context) (*godo.Firewall, error) {
+	// check cache and query the API firewall service to get firewall ID, if it exists. Return it. If not, continue.
+	if fm.fwCache.firewall != nil {
+		fm.fwCache.mu.RLock()
+		fwID := fm.fwCache.firewall.ID
+		fm.fwCache.mu.RUnlock()
+
+		fw, resp, err := fm.client.Firewalls.Get(ctx, fwID)
+		if err != nil && (resp == nil || resp.StatusCode != http.StatusNotFound) {
+			return nil, fmt.Errorf("failed to retrieve firewall by ID: %s", err)
+		}
+		if resp.StatusCode == http.StatusNotFound {
+			klog.Warningf("unable to retrieve firewall by ID because it no longer exists")
+		}
+		if fw != nil {
+			return fw, nil
+		}
+	}
+
+	// iterate through firewall API provided list and return the firewall with the matching firewall name.
+	f := func(fw godo.Firewall) bool {
+		return fw.Name == fm.workerFirewallName
+	}
+	klog.Info("filtering firewall list for the firewall that has the expected firewall name")
+	fw, err := filterFirewallList(ctx, fm.client, f)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve list of firewalls from DO API: %s", err)
+	}
+	return fw, nil
+}
+
+// Set applies the given inbound rules to the public access firewall when the current rules and target rules differ.
+func (fm *firewallManagerOp) Set(ctx context.Context, svcInboundRules []godo.InboundRule) error {
+	fm.fwCache.mu.RLock()
+	targetFirewall := fm.fwCache.firewall
+	isEqual := false
+	if targetFirewall != nil && cmp.Equal(targetFirewall.InboundRules, svcInboundRules) {
+		isEqual = true
+	}
+	fm.fwCache.mu.RUnlock()
+
+	if isEqual {
+		return nil
+	}
+
+	fwSvc := fm.client.Firewalls
+
+	// Check if the target firewall ID exists but in the case that CCM first starts up and the
+	// firewall ID does not exist yet, we have to retrieve one from the API.
+	var fwID string
+	if targetFirewall == nil {
+		currentFirewall, err := fm.Get(ctx)
+		if err == nil && currentFirewall == nil {
+			// Log error then create firewall and update cache, since firewall does not exist.
+			klog.Errorf("failed to get firewall state: %s", err)
+			currentFirewall, err = fm.createFirewall(ctx, fwSvc, svcInboundRules)
```

Comment:
If the `createFirewall` call succeeds, we still invoke the update below which we shouldn't need to do anymore. We can duplicate the final cache update and return prematurely.

Reply:
Are you suggesting to remove?:
```go
currentFirewall, resp, err := fwSvc.Update(ctx, fwID, fr)
```

So that this is no longer there:
```go
// Update api firewall inbound rules with the expected inbound rules then update cache.
fr := &godo.FirewallRequest{
	Name:         fm.workerFirewallName,
	InboundRules: svcInboundRules,
	Tags:         fm.workerFirewallTags,
}

// Attempt to update the firewall service with the new inbound rules.
currentFirewall, resp, err := fwSvc.Update(ctx, fwID, fr)
if err != nil {
	if resp == nil || resp.StatusCode != http.StatusNotFound {
		return fmt.Errorf("failed to update firewall: %v", err)
	}
	// Firewall does not exist anymore, so we need to create a new firewall with the
	// updated inbound rules.
	currentFirewall, err = fm.createFirewall(ctx, fwSvc, svcInboundRules)
	if err != nil {
		return fmt.Errorf("failed to add new firewall inbound rules: %s", err)
	}
}
```

The **first check**, at the very top, checks to see if the cache is nil AND that the inboundRules are equal:
```go
fm.fwCache.mu.RLock()
targetFirewall := fm.fwCache.firewall
isEqual := false
if targetFirewall != nil && cmp.Equal(targetFirewall.InboundRules, svcInboundRules) {
	isEqual = true
}
fm.fwCache.mu.RUnlock()

if isEqual {
	return nil
}
```

The **next check** is for when `targetFirewall == nil`. We create a firewall (and I agree with your suggestion, that we should update the cache then return early"

With that being said, should we also keep the **final check** which _should_ atleast happen when `targetFirewall != nil` so that the rules we want are what we expect, both from the firewall service and the cache? Which I think can be achieved by leaving the update logic in but making sure to return after the first successful call to `createFirewall`.

Wdyt?